### PR TITLE
Beginning of (backards-compatible) Key pickling

### PR DESCRIPTION
**Pull request by jquast**
This isn't quite ready, I need to add test coverage for migrating old databases, esp. around the conditionals in ``__delitem__`` and ``__contains__`` by manually executing db.conn.execute() statements to add "old-style bytes".  Let me know what y'all think anyway, @ziky90, @piskvorky

Details of the implementation can be found in the comments of methods ``decode_key``, ``SqliteDict.__keysearch``, and ``SqliteDict.__getitem__`` about the v1.2- and v2.0+ differences.

I covered some of the more complex problems, such as, when a key is set in v1.2 using unicode, then later deleted also using unicode:
```
>>> db[u'unicode'] = 'stuff'    # v1.2
>>> db[u'unicode'] = 'rewritten' # v2.0
```

That the original utf-8 encoded b'unicode' gets deleted after the new pickled-unicode key u'unicode' is inserted.  Some careful test coverage about transparently migrating old database keys into new ones is required.  In this scenario, db.keys() would contain b'unicode' if it was written by v1.2, but after a rewrite with v2.0, then db.keys() would contain u'unicode', which might be tricky enough to warrant a mention in the README.

I'm afraid if I say too much more it will only confuse things :)

Development process was recorded:

- https://www.livecoding.tv/video/foss-gardening-sqlitedict-8/
- https://www.livecoding.tv/video/foss-gardening-sqlitedict-9/

### CODE REVIEW

**Comment by ziky90** [comment]
I am closing the PR #26.
I have just removed unused test file with just comments and done some newlines changes.
I just went through tests and code and it seems good to me, though I agree with required tricky tests dealing with the migration.

File: sqlitedict.py [hunk 1]
```diff
@@ -180,40 +221,120 @@ def __bool__(self):
...
+            yield key
+
+        # for python 2.x, if the given key is of type 'unicode',
+        # then search for the utf-8 encoded form (v1.2-)
+        elif _major_version < 3:
+            if isinstance(key, unicode):
+                try:
+                    yield key.encode('utf8')
+                except UnicodeEncodeError:  # unlikely
+                    pass
```
Comment (ziky90) [inline sqlitedict.py]: But still I'd add here at least something like `logger.warning("UnicodeEncodeError")`, etc.

File: sqlitedict.py [hunk 2]
```diff
@@ -180,40 +221,120 @@ def __bool__(self):
...
+
+        # and for python 3.x, 'str' is actually the new 'unicode' type,
+        # perform the same coercion search.
+        elif isinstance(key, str):
+            try:
+                yield key.encode('utf8')
+            except UnicodeEncodeError:  # unlikely
+                pass
```
Comment (ziky90) [inline sqlitedict.py]: The same here.

**Comment by piskvorky** [comment]
Thanks Jeff!

If the "auto-migration" code turns out to be too hairy, I'm +1 on offering an explicit one-off migration script to users. And keeping the library code lean and clean (pickle-only) -- after all, the goal of `sqlitedict` is to be simple and lightweight.

Just throwing the option out there -- I leave it to your judgement whether we've crossed the line of "too much complexity".

**Comment by jquast** [comment]
I do think we crossed "too much complexity", couldn't quite say so until the code was written :)

I'm thinking of moving this code into a single method/procedure, that auto-upgrades databases on first-open, only, perhaps making a .v12.backup file as well ?

**Comment by ziky90** [comment]
From my point of view, more code dealing with backward compatibility than actual logic is not good. But auto upgrade would probably make code easier to read and understand.

But still I'd be better for breaking the backward compatibility and for forcing users to migrate the sqlite explicitly by separate python script to the version compatible with sqlitedict 2.x. I can imagine that there might be also other applications using the same sqlite DB than just sqlitedict and automatic migration also might cause problems here (this partially solves the .v12.backup, that's true).

**Comment by jquast** [comment]
> But still I'd be better for breaking the backward compatibility and for forcing users to migrate the sqlite explicitly by separate python script

I have software that uses sqlitedict as a backend -- the users you would force in such a case were not even aware of sqlitedict until it breaks! As long as the script is distributed and accessible as a module, I could always write "auto-upgrade" process downstream along with being able to use unicode as original intended.

My largest fear is the common mistake of people adding "sqlitedict" only as a requirement in setup.py or requirements.txt -- their package suddenly breaks against their existing data or code that assumes bytes or py2.7, not unicode.  Even worse, "users" may pin their direct dependencies, but the transient dependencies they require may not freeze, either.

The more involved it gets, the more in favor of the original proposal https://github.com/piskvorky/sqlitedict/issues/19#issuecomment-67394915 of simply distributing a new python package all together (eg. sqlitedict2) :)

**Comment by ziky90** [comment]
>
My largest fear is the common mistake of people adding "sqlitedict" only as a requirement in setup.py or requirements.txt -- their package suddenly breaks against their existing data or code that assumes bytes or py2.7, not unicode. Even worse, "users" may pin their direct dependencies, but the transient dependencies they require may not freeze, either.
>

I agree with this. I usually try to have `package==version` in my requirements.txt files, but I agree that not everyone does this.

>
The more involved it gets, the more in favor of the original proposal #19 (comment) of simply distributing a new python package all together (eg. sqlitedict2) :)
>

I also think that this might be a good solution. They've done this for example in the boto library with boto3 https://github.com/boto/boto#boto-3. What do you think? @piskvorky 
Old projects will simply stay with sqlitedict and new projects will be developed with sqlitedict2 + sqlitedict2 might contain the DB migration script as well.

**Comment by piskvorky** [comment]
I'm mildly -1 on a new package. It doesn't feel right to fragment the project... Would we then backport fixes too?

An explicit **migration script** `convert_sqlitedict1to2.py olddb.sqlite newdb.sqlite` + a loud **explicit Exception** whenever users try to use old DB with new code (or vice versa) sounds the cleanest.

I think this is @ziky90 's proposal in #30 .

How much trouble would such upgrade process be in your use case, @jquast ?

We cannot cover 100% use cases, we don't even know what they are. But making it easy to migrate for the most common cases is fine by me. Plus clear exceptions, fat doc warnings and a README section, of course.

**Comment by jquast** [comment]
We definitely differ on the package split. I can think of too many edge cases of destructively corrupting dependent's data that I wouldn't want my name affiliated with the the release that causes it, that's how bad I think it is.

**Comment by jquast** [comment]
Rejected, use #30.

**Comment by jquast** [comment]
Some examples of trouble -- its not just migrating the data, it is also migrating the code that reads or writes the data.

https://github.com/jquast/x84/blob/ab8551896490108a960bdcbd07059b75c2a7b335/x84/bbs/msgbase.py#L56
https://github.com/jquast/x84/blob/ab8551896490108a960bdcbd07059b75c2a7b335/x84/bbs/msgbase.py#L67
https://github.com/jquast/x84/blob/ab8551896490108a960bdcbd07059b75c2a7b335/x84/bbs/msgbase.py#L81
https://github.com/jquast/x84/blob/ab8551896490108a960bdcbd07059b75c2a7b335/x84/bbs/msgbase.py#L161
https://github.com/jquast/x84/blob/ab8551896490108a960bdcbd07059b75c2a7b335/x84/bbs/userbase.py#L17-18
https://github.com/jquast/x84/blob/ab8551896490108a960bdcbd07059b75c2a7b335/x84/bbs/userbase.py#L39
https://github.com/jquast/x84/blob/ab8551896490108a960bdcbd07059b75c2a7b335/x84/default/msgarea.py#L506
https://github.com/jquast/x84/blob/6340573b68f4d4dcdf860795b433ba541164686f/x84/default/tetris.py#L65
https://github.com/jquast/x84/blob/ab8551896490108a960bdcbd07059b75c2a7b335/x84/default/sysop.py#L71

This is mostly on the "reading" end of the pipe, the writing end of the pipe might have to be changed to.  A simple database upgrade script would do nothing here: the database would be corrupted immediately after its first write, and any subsequent upgrade script would "pickle an already pickled string", further corrupting. I do not want to be associated with this next sqlitedict release, I will not be responsible for corrupting user data.

**Comment by piskvorky** [comment]
I see you feel strongly about this. Let's work it out, nothing's final.

The change of "keys are now pickled" is meant to be transparent from code POV (after data migration) -- why do you think it isn't? Under what circumstance will you need to change your code?

I'm not sure what I'm supposed to be looking at in these x84 links, I'm not familiar with that project. Calling `.keys()` or `del` should work exactly the same way before and after, right @ziky90 ?

Or are you mixing "sqlitedict access" to a sqlite DB with "direct access" to the same DB file? Is that it?

**Comment by jquast** [comment]
```
db["start"] = True
db.commit()
db.get("last_stopped", None) raises KeyError
```

I'm now in the situation where i have one pickled key, "start", and all my remaining keys unpickled. Running the migration script would only "pickle a pickle". You can't depend on decode() catching the error before a user has already begun to write into a v1.2 database with v2.0 data.

**Comment by jquast** [comment]
Also you're looking at things like coercing bytes back to unicode, "ascii numbers" back to real numbers, etc.  This is because of their implicit coercion to ascii bytes -- upgrading the database would break such code.

**Comment by ziky90** [comment]
>
Calling .keys() or del should work exactly the same way before and after, right @ziky90 ?
>

.keys() would try to call `decode()` since version 2.x. on every key and del would be deleting the element based on the encoded key. 

>
I'm now in the situation where i have one pickled key, "start", and all my remaining keys unpickled. Running the migration script would only "pickle a pickle". You can't depend on decode() catching the error before a user has already begun to write into a v1.2 database with v2.0 data.
>

I agree, this might be a serious problem. Having part of the DB with pickled keys and half with plain string keys would make it impossible to use migration script. Seeing this I am starting to like the idea of adding pickling with just migration script less and less.

**Comment by piskvorky** [comment]
Hold on, I'm getting lost. Let me summarize. We're considering two options:

1. we release "keys are pickled" as current package sqlitedict, version 2.0
2. we release "keys are pickled" as a new package sqlitedict**2**, version 2.0

Either way, users have to migrate their existing database for the new version. The migration will be done by an explicit CLI script, call signature `migrate_sqlitedict1to2.py old_db.sqlite new_db.sqlite`.

Issue (both versions): users who

a) upgrade between 1.x and 2.x
b) AND don't migrate data explicitly (don't know about, or care, about the major version change),
c) AND have a pre-existing database, 

will get a mix of pickled and non-pickled keys in their database. Broken.

Difference between option 1. and 2. is that 2. better avoids "accidental" upgrades: user doesn't have the version of sqlitedict fixed in requirements, does a) unknowningly.

Either way, I didn't get the part around why code would have to be changed @jquast . I believe user code stays the same between version, only data needs to be migrated, right? Old version = keys are byte strings; new version = keys are byte strings. No difference in types, only in interpretation.

**Comment by jquast** [comment]
Yes, I think you understand the 1/2/a/b/c correctly.

> I didn't get the part around why code would have to be changed

Writing an examining key types with v1:

```
>>> db = sqlitedict.SqliteDict('./test.sqld', tablename='unicode')
>>> db[u"Řehůřek"] = 1
>>> for key in db:
....     print(repr(key))
....
'\xc5\x98eh\xc5\xaf\xc5\x99ek'

>>> db = sqlitedict.SqliteDict('./test.sqld', tablename='ints')
>>> db[1] = 1
>>> db[2] = 2
>>> for key in db:
....     print(repr(key))
....
'1'
'2'
```

In both cases, to pull the data from the database would require some code:

```
>>> db = sqlitedict.SqliteDict('./test.sqld', tablename='unicode')
>>> for key in (_key.decode('utf8') for _key in db):
....     print(repr(key))
u'\u0158eh\u016f\u0159ek'
>>> db = sqlitedict.SqliteDict('./test.sqld', tablename='ints')
>>> for key in (int(_key) for _key in db):
....     print(repr(key))
1
2
```

In the first case, upgrading to sqlitedict v2 would fail: the code continues to insert unicode, and decode it on its way out, but it is already unicode on its way out:

```
>>> u'\u0158eh\u016f\u0159ek'.decode('utf8')
UnicodeEncodeError: 'ascii' codec can't encode character u'\u0158' in position 0: ordinal not in range(128)
```

In the second case, you are just lucky that int(int) == int :)

**Comment by piskvorky** [comment]
Ah, I see what you mean.

Non-string keys were never supported, so backward compatibility there is no issue.

In other words, if users relied on undocumented behaviour (passing keys as unicode, or int, or whatever), they'll have to fix their code. Only string keys are officially supported in v1.

If this is what's happening in your repo Jeff, I think you can't upgrade to v2.0 anyway, regardless of whether we go with option 1 (major version only) or option 2 (major + new package). Neither option will work with code that expects implicit type coercions done by sqlitedict.

**Comment by piskvorky** [comment]
@jquast Released 1.3.0 (your fixes around handling threading errors; no key pickling there).

I'll probably release 2.0 as a separate package next week, like you advised.

In any case, your old code likely won't work in 2.0 (if you store number keys but expect string keys back, the code will break, see my comment above).

**Comment by jquast** [comment]
great

just to be clear I'm not really concerned with my code's compatibility, but the unknown number of other's. I version pin, of course :) thanks for all your consideration!

**Comment by piskvorky** [comment]
@jquast release coming up (work interfered)!

Should I revive this branch or what is the PR we ultimately decided to use? It's not #30 -- is it #31?

**Pull request closed without merging**